### PR TITLE
[WIP] EC FW sync

### PR DESCRIPTION
**Pull request by mkopec**
Change-Id: I7f2c9f756102ad6e63d8aafd3013003d31676ce4

### CODE REVIEW

**Comment by mkopec** [comment]
So far it can compare versions:

```
[DEBUG]  EC update found (46113 bytes)
[DEBUG]  EC update target:  clevo/nv40mz
[DEBUG]  EC update version:  2023-12-07_bf0073c
[DEBUG]  EC current version: 2023-12-08_2b2c17a
[DEBUG]  EC update required!
```

File: src/ec/system76/ec/system76_ec.c [hunk 1]
```diff
@@ -148,3 +187,156 @@ void system76_ec_set_bat_threshold(enum bat_threshold_type type, uint8_t value)
...
+		0,
+	};
+
+	/* If we failed to jump to scratch ROM, then we can probably continue booting. */
+	if (system76_ec_smfi_cmd(CMD_SPI, sizeof(smfi_cmd) / sizeof(uint8_t), (uint8_t *)&smfi_cmd)) {
+		printk(BIOS_ERR, "EC update failed!\n");
+		return;
+	}
+
+	uint8_t current_rom[SPI_FLASH_SIZE] = { 0xFF };
```
Comment (miczyg1) [inline src/ec/system76/ec/system76_ec.c:328]: Don't do it like that. You basically request 128K of memory to be allocated on stack. coreboot's stake is only a couple of kilobytes...

Do a malloc instead and free the buffer after use. Then the buffer will be used as heap.

@mkopec

Comment (mkopec) [reply src/ec/system76/ec/system76_ec.c]: makes sense, applied your suggestion in https://github.com/Dasharo/coreboot/pull/456/commits/e34fee6c8657ee449fdf5824651ad457f7c81f01 but it fails as well with this error:

```
memalign(boundary=8, size=131072): failed: Tried to round up free_mem_ptr 0x76b0ba30 to 0x76b2ba30
but free_mem_end_ptr is 0x76b1ba30
```

I suppose we don't actually need to allocate the full 128KB, for verification after update we can just compare chunk by chunk

Comment (miczyg1) [reply src/ec/system76/ec/system76_ec.c]: As you can see, coreboot heap is 64KiB too short to allocate full flash size.

> I suppose we don't actually need to allocate the full 128KB, for verification after update we can just compare chunk by chunk

Probably better idea.

**Pull request closed without merging**